src/client/dfuse/pil4dfs/int_dfs.c

```diff
@@ -3247,6 +3255,166 @@ closedir(DIR *dirp)
 	}
 }
 
+long
+telldir(DIR *dirp)
+{
+	int fd;
+
+	if (next_telldir == NULL) {
+		next_telldir = dlsym(RTLD_NEXT, "telldir");
+		D_ASSERT(next_telldir != NULL);
+	}
+	if (!hook_enabled)
+		return next_telldir(dirp);
+
+	fd = dirfd(dirp);
+	if (fd < FD_DIR_BASE)
+		return next_telldir(dirp);
+
+	return dir_list[fd - FD_DIR_BASE]->offset;
+}
+
+void
+rewinddir(DIR *dirp)
+{
+	int fd, idx;
+
+	if (next_rewinddir == NULL) {
+		next_rewinddir = dlsym(RTLD_NEXT, "rewinddir");
+		D_ASSERT(next_rewinddir != NULL);
+	}
+	if (!hook_enabled)
+		return next_rewinddir(dirp);
+
+	fd = dirfd(dirp);
+	if (fd < FD_DIR_BASE)
+		return next_rewinddir(dirp);
+
+	idx                     = fd - FD_DIR_BASE;
+	dir_list[idx]->offset   = 0;
+	dir_list[idx]->num_ents = 0;
+	memset(&dir_list[idx]->anchor, 0, sizeof(daos_anchor_t));
+
+	return;
+}
+
+/* Offset of the first entry, allow two entries for . and .. */
+#define OFFSET_BASE 2
+
+void
+seekdir(DIR *dirp, long loc)
+{
+	int      fd, idx, rc;
+	long     num_entry;
+	uint32_t num_to_read;
+
+	if (next_seekdir == NULL) {
+		next_seekdir = dlsym(RTLD_NEXT, "seekdir");
+		D_ASSERT(next_seekdir != NULL);
+	}
+	if (!hook_enabled)
+		return next_seekdir(dirp, loc);
+
+	fd = dirfd(dirp);
+	if (fd < FD_DIR_BASE)
+		return next_seekdir(dirp, loc);
+
+	idx = fd - FD_DIR_BASE;
+
+	/* need to compare loc with current offset & the number of cached entries */
+	if (loc <= OFFSET_BASE) {
+		dir_list[idx]->offset   = loc;
+		dir_list[idx]->num_ents = 0;
+		memset(&dir_list[idx]->anchor, 0, sizeof(daos_anchor_t));
+		return;
+	}
+	if (dir_list[idx]->offset <= OFFSET_BASE) {
+		/* no buffered entry */
+		dir_list[idx]->offset   = OFFSET_BASE;
+		dir_list[idx]->num_ents = 0;
+		num_entry               = loc - OFFSET_BASE;
+	} else if (loc < dir_list[idx]->offset) {
+		/* rewind and read entries from the beginning */
+		dir_list[idx]->offset   = OFFSET_BASE;
+		dir_list[idx]->num_ents = 0;
+		memset(&dir_list[idx]->anchor, 0, sizeof(daos_anchor_t));
+		num_entry = loc - dir_list[idx]->offset;
+	} else if (loc >= (dir_list[idx]->offset + dir_list[idx]->num_ents)) {
+		/* need to read more entries from current offset */
+		dir_list[idx]->offset   = dir_list[idx]->offset + dir_list[idx]->num_ents;
+		dir_list[idx]->num_ents = 0;
+		num_entry               = loc - dir_list[idx]->offset;
+	} else if (loc >= dir_list[idx]->offset) {
+		/* in the cached entries */
+		dir_list[idx]->num_ents -= (loc - dir_list[idx]->offset);
+		dir_list[idx]->offset = loc;
+		return;
+	}
+
+	while (num_entry) {
+		num_to_read = min(READ_DIR_BATCH_SIZE, num_entry);
+		rc          = dfs_iterate(dir_list[idx]->dfs_mt->dfs, dir_list[idx]->dir,
+					  &dir_list[idx]->anchor, &num_to_read,
+					  DFS_MAX_NAME * num_to_read, NULL, NULL);
```

Comment:
FYI - if you don't like this style (and I don't) then leave a blank line between the assignment and the function call, then clang-format will flow it differently which should make it more readable.
```suggestion
		num_to_read = min(READ_DIR_BATCH_SIZE, num_entry);

		rc = dfs_iterate(dir_list[idx]->dfs_mt->dfs, dir_list[idx]->dir, &dir_list[idx]->anchor,
					  &num_to_read, DFS_MAX_NAME * num_to_read, NULL, NULL);
```

Reply:
Thank you! Good to know it. I just reformatted it.